### PR TITLE
Improve performance

### PR DESCRIPTION
**Pull request by tobiasdiez**
Improve performance by:
- Caching fields as KeywordList
- Remove duplicate change listener
- Remove preference access in loop

With these changes, the mega huge library of @AEgit works without too much delays (the initial loading of the database is a bit slow but still ok). Thus, this PR is a potential fix for #4430 and for https://github.com/JabRef/jabref/issues/5071.

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: src/main/java/org/jabref/gui/BasePanel.java [hunk 1]
```diff
@@ -908,10 +908,6 @@ public void updateEntryEditorIfShowing() {
 
     public void markBaseChanged() {
         baseChanged = true;
-        markBasedChangedInternal();
-    }
-
-    private void markBasedChangedInternal() {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/BasePanel.java]: I remember hard discussions when introducing this. That we really need it etc. What a pity that is was not documented. In retrospective, that would have been interesting.

Nice to see that this special function is gone now.

File: src/main/java/org/jabref/gui/BasePanel.java [hunk 2]
```diff
@@ -1085,11 +1081,6 @@ public void cut() {
         mainTable.cut();
     }
 
-    @Subscribe
-    public void listen(EntryChangedEvent entryChangedEvent) {
-        this.markBaseChanged();
```
Comment (koppor) [inline src/main/java/org/jabref/gui/BasePanel.java]: Interesting that this is not needed anymore. - I tested it locally and the star works fine.

File: src/main/java/org/jabref/model/entry/BibEntry.java [hunk 3]
```diff
@@ -56,6 +57,11 @@
      * Cache that stores latex free versions of fields.
      */
     private final Map<Field, String> latexFreeFields = new ConcurrentHashMap<>();
+    /**
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/BibEntry.java]: add an empty line before 😇

File: src/main/java/org/jabref/model/entry/BibEntry.java [hunk 4]
```diff
@@ -825,33 +830,46 @@ public String getUserComments() {
         }
     }
 
+    public KeywordList getFieldAsKeywords(Field field, Character keywordSeparator) {
+        Optional<KeywordList> storedList = fieldsAsKeywords.get(field, keywordSeparator);
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/BibEntry.java]: Can't you use [Google Guava's Cache](https://www.baeldung.com/guava-cacheloader)? 😇  Maybe, with [Eclipse Collections Twin](https://www.eclipse.org/collections/javadoc/9.1.0/org/eclipse/collections/api/tuple/Twin.html)

```java
LoadingCache<Pair<Field, Character>, KeywordList> storedKeywordList = CacheBuilder.newBuilder()
  .build(new CacheLoader<Pair<Field, Character>, KeywordList>() {
    @Override
    public String load(final Pair<Field, Character> pair) throws Exception {
      return getField(pair.getOne().map(... pair.getTwo());
    }
});
```

```java
return storedKeywordList.get(Tuples.pair(field, keywordSeparator));
```

OK, invalidation is worse (as it is not possible to invalidate all pairs having the same first value (easily)).

```java
storedKeywordList.invalidateAll();
```

(I now that Tuples.pair(...) might be slower than your approach. Howver, I would like that the cache can be tweaked ^^).

File: src/main/java/org/jabref/model/entry/BibEntry.java [hunk 5]
```diff
@@ -825,33 +830,46 @@ public String getUserComments() {
...
-            Optional<String> citeKey = getCiteKeyOptional();
-            latexFreeFields.put(field, citeKey.get());
-            return citeKey;
+            return getCiteKeyOptional();
         } else if (InternalField.TYPE_HEADER.equals(field)) {
-            String typeName = type.get().getDisplayName();
-            latexFreeFields.put(field, typeName);
-            return Optional.of(typeName);
+            return Optional.of(type.get().getDisplayName());
+        } else if (latexFreeFields.containsKey(field)) {
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/BibEntry.java]: This could also be a good place for Guava's Cache.

**Comment by AEgit** [comment]
@tobiasdiez Wonderful! Let me know, when I can start testing this fix.

File: src/main/java/org/jabref/gui/BasePanel.java (hunk 2 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/BasePanel.java]: EntryChangedEvents are included in the DatabaseChangedEvents (which also trigger a `markBaseChanged`)

File: src/main/java/org/jabref/model/entry/BibEntry.java (hunk 4 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/model/entry/BibEntry.java]: That looks complicated 😄 https://github.com/google/guava/wiki/CachesExplained#applicability has the nice explanation
> Note: If you do not need the features of a Cache, ConcurrentHashMap is more memory-efficient -- but it is extremely difficult or impossible to duplicate most Cache features with any old ConcurrentMap.

I guess we do not need these cache features (but I've also never worked with Guave Caches...so I cannot appreciate them).

**Pull request merged**